### ISSUE TITLE
[Bug Report] TransformerBridge.load_state_dict(assign=True) desyncs tied embed/unembed weights, silently breaking forward pass

### ISSUE THREAD

**Issue by LightWork666** [TransformerLensOrg/TransformerLens#1725]
**Describe the bug**

GPT-2 (and any architecture with `tie_word_embeddings`) ties its embedding and unembedding weights -- `transformer.wte.weight` and `lm_head.weight` share the same underlying storage even though, in the bridge-wrapped model, they are already separate `nn.Parameter` objects pointing at that shared storage (confirmed via `untyped_storage().data_ptr()` equality at construction, before any load -- structurally the same "different object, shared storage" pattern as #1637's split QKV components, just full-size rather than a partial view).

`load_state_dict(state_dict, assign=True)` breaks this the same way it broke split components in #1637: PyTorch's `assign=True` replaces each target parameter object independently per key, so loading only one of the two tied keys (e.g. just `embed.weight`, a completely ordinary partial/fine-tuned checkpoint) leaves the other holding stale pre-load data with no error or warning.

This is worse than #1637 in one respect: there, the bridge's own forward pass stayed correct and only the raw export (`original_model.state_dict()` / `save_pretrained()`) went stale. Here, `UnembeddingBridge.forward()` reads `original_component.weight` directly -- the now-desynced, stale `lm_head` weight -- so **the bridge's own live forward pass silently uses the wrong weights**, not just its export.

Also distinct from #1637 mechanically: this is not a partial-view case, so the `_is_view_backed` detection added in #1660 does not (and structurally cannot) catch it -- both tensors are full-size, `_is_view_backed` returns `False` for both.

**Code example**

```python
import torch
from transformer_lens.model_bridge import TransformerBridge

bridge = TransformerBridge.boot_transformers("gpt2", device="cpu")
bridge.eval()
tokens = torch.randint(0, 1000, (1, 6))
with torch.no_grad():
    logits_before = bridge(tokens).clone()

sd = {k: v.clone() for k, v in bridge.state_dict().items()}
new_embed = sd["embed.weight"] + 3.0
bridge.load_state_dict({"embed.weight": new_embed}, strict=False, assign=True)

with torch.no_grad():
    logits_after = bridge(tokens).clone()

print(torch.allclose(logits_before, logits_after))  # -> True: the load was a silent no-op on the forward pass
print(torch.allclose(bridge.unembed.original_component.weight, sd["embed.weight"]))  # -> True: unembed is still the OLD embed value
```

No error, no warning. The embedding matrix was successfully replaced, but every downstream computation through the unembedding silently keeps using the pre-load weights, because the tie desynced.

Separately, even a *complete* round-trip load (both `embed.weight` and `unembed.weight` present with consistent values, so no data loss) still silently breaks the tie at the object level -- doubling memory for that weight matrix and decoupling their gradients for any future training, with no error either.

`assign=False` (the default) is unaffected in both cases -- verified the tie and forward pass stay correct.

**System Info**

- Installed from source, `dev-4.x`, on top of #1660
- OS: macOS (arm64), CPU-only
- Reproduces with `boot_transformers("gpt2")`; mechanism applies to any architecture with `tie_word_embeddings=True`

**Expected behaviour & fix pointers**

Same family of fix as #1637 (option 3 + option 1: route `assign=True` for these keys through explicit in-place `.copy_()`, raise a clear error on shape mismatch), but the *detection* needs to change -- `_is_view_backed`'s "smaller than its own storage" check is specific to partial views and doesn't fire for a full-size tensor that merely shares storage with another same-size tensor. Detecting this case needs something like: for each incoming key, check whether any *other* current parameter/buffer shares storage with this key's current target (a storage-identity index built from `original_model.state_dict(keep_vars=True)`), not just a self-contained size check.

Acceptance:
- [ ] Loading only `embed.weight` (or only `unembed.weight`) with `assign=True` does not leave the other stale in either `original_model.state_dict()` or the bridge's own forward pass
- [ ] A complete tied-pair load with `assign=True` does not silently break the storage tie
- [ ] `assign=False` and existing #1587/#1615/#1637 tests are unaffected
- [ ] New regression test on real gpt2 (tied by default) exercising both the partial-load and full-round-trip cases
- [ ] `make unit-test` and `uv run mypy .` pass

### Checklist

- [x] I have checked that there is no similar issue in the repo (**required**)
